File: src/cursors.mjs
// Cursor store for mutable byte streams and immutable closed segments. Legacy
// records retain their numeric coercion contract; tagged segment commits are
// stored as structured records and never projected onto byte-offset fields.
// Writes are durable: unique temp, file sync, rename, then parent sync.
// No digit characters outside quoted strings and comments.
import { spawnSync } from 'node:child_process';
import { closeSync, existsSync, fsyncSync, mkdirSync, openSync, readFileSync, renameSync, rmSync, unlinkSync, writeFileSync } from 'node:fs';
import { randomUUID } from 'node:crypto';
import { hostname } from 'node:os';
import { join } from 'node:path';

const N = (s) => Number(s);
const ZERO = N('0');
const TWO = N('2');
const SEGMENT_KIND = 'closed-segment';

function coerce(rec) {
  const mtimeMs = Number(rec.mtimeMs);
  const size = Number(rec.size);
  const offset = Number(rec.offset);
  if (
    !Number.isFinite(mtimeMs)
    || !Number.isFinite(size)
    || !Number.isFinite(offset)
    || size < ZERO
    || offset < ZERO
  ) {
    throw new Error('cursor record contains invalid numeric state');
  }
  return {
    mtimeMs,
    size,
    offset,
  };
}

// Cursor loss can replay already-persisted evidence, so unreadable state is a
// hard failure. Recovery uses a separate empty LAKE_DATA root, never a silent
// fallback that appends a second copy to existing partitions.
function readStore(filePath) {
  if (!existsSync(filePath)) return {};
  let raw;
  try {
    raw = readFileSync(filePath, 'utf8');
  } catch (error) {
    throw new Error('cursor store is unreadable; preserve it and recover into an empty LAKE_DATA: ' + String(error));
  }
  let parsed;
  try {
    parsed = JSON.parse(raw);
  } catch (error) {
    throw new Error('cursor store is corrupt; preserve it and recover into an empty LAKE_DATA: ' + String(error));
  }
  if (!parsed || typeof parsed !== 'object' || Array.isArray(parsed)) {
    throw new Error('cursor store must be a JSON object; preserve it and recover into an empty LAKE_DATA');
  }
  return parsed;
}
function syncDirectory(path) {
  const fd = openSync(path, 'r');
  try { fsyncSync(fd); } finally { closeSync(fd); }
}

function processIdentity(pid) {
  const result = spawnSync('/bin/ps', ['-o', 'lstart=', '-p', String(pid)], { encoding: 'utf8' });
  if (result.error || result.status !== ZERO) return null;
  return result.stdout.trim() || null;
}

function ownerIsAlive(owner) {
  if (!owner || !Number.isInteger(owner.pid) || owner.pid <= ZERO) return false;
  if (owner.host !== hostname()) return true;
  const currentIdentity = processIdentity(owner.pid);
  if (owner.started && currentIdentity) return owner.started === currentIdentity;
  try {
    process.kill(owner.pid, ZERO);
    process.stderr.write('cursors: lock process identity unavailable; retaining live pid claim\n');
    return true;
  } catch (error) {
    return !error || error.code !== 'ESRCH';
  }
}

// Build and sync each claim privately, then publish it with one rename. This
// makes an absent or malformed published owner structurally abandoned rather
// than a live process caught between mkdir and writing its identity.
function acquireLock(dataDir, lockPath) {
  const token = randomUUID();
  const owner = { host: hostname(), pid: process.pid, started: processIdentity(process.pid), token };
  for (;;) {
    const prepared = lockPath + '.claim-' + process.pid + '-' + randomUUID();
    mkdirSync(prepared);
    let fd = null;
    try {
      fd = openSync(join(prepared, 'owner.json'), 'wx');
      writeFileSync(fd, JSON.stringify(owner));
      fsyncSync(fd);
      closeSync(fd);
      fd = null;
      syncDirectory(prepared);
      renameSync(prepared, lockPath);
      syncDirectory(dataDir);
      return token;
    } catch (error) {
      if (fd !== null) closeSync(fd);
      if (existsSync(prepared)) rmSync(prepared, { recursive: true, force: true });
      if (!error || (error.code !== 'EEXIST' && error.code !== 'ENOTEMPTY')) throw error;
    }

    let incumbent = null;
    try { incumbent = JSON.parse(readFileSync(join(lockPath, 'owner.json'), 'utf8')); } catch {}
    if (ownerIsAlive(incumbent)) {
      throw new Error(
        'state writer lock is held by '
        + String(incumbent.host || 'unknown-host')
        + ' pid ' + String(incumbent.pid || 'unknown')
      );
    }
    try { rmSync(lockPath, { recursive: true }); } catch (error) {
      if (!error || error.code !== 'ENOENT') throw error;
    }
  }
}

function releaseLock(dataDir, lockPath, token) {
  let owner = null;
  try { owner = JSON.parse(readFileSync(join(lockPath, 'owner.json'), 'utf8')); } catch {}
  if (!owner || owner.token !== token) return;
  rmSync(lockPath, { recursive: true });
  syncDirectory(dataDir);
}

export function openWriterLease(dataDir) {
  mkdirSync(dataDir, { recursive: true });
  const lockPath = join(dataDir, 'ingest.lock');
  const token = acquireLock(dataDir, lockPath);
  let closed = false;
  return {
    close() {
      if (closed) return;
      releaseLock(dataDir, lockPath, token);
      closed = true;
    },
  };
}

export function openCursors(dataDir) {
  const filePath = join(dataDir, 'cursors.json');
  const lockPath = join(dataDir, 'cursors.lock');
  let store = readStore(filePath);
  const pending = new Map();
  let dirty = false;

  function get(file) {
    const rec = store[file];
    if (!rec || typeof rec !== 'object') return null;
    if (rec.kind === SEGMENT_KIND) return structuredClone(rec);
    return coerce(rec);
  }

  function set(file, rec) {
    if (!rec || typeof rec !== 'object') return;
    const value = rec.kind === SEGMENT_KIND ? structuredClone(rec) : coerce(rec);
    store[file] = value;
    pending.set(file, value);
    dirty = true;
  }

  function flush() {
    if (!dirty) return;
    mkdirSync(dataDir, { recursive: true });
    const token = acquireLock(dataDir, lockPath);
    const tmpPath = filePath + '.tmp-' + process.pid + '-' + randomUUID();
    let fd = null;
    try {
      // Reload only after taking ownership: the lock protects this entire
      // read-modify-write transaction, not merely the final rename.
      const merged = readStore(filePath);
      for (const [file, value] of pending) merged[file] = value;
      fd = openSync(tmpPath, 'wx');
      writeFileSync(fd, JSON.stringify(merged, null, TWO));
      fsyncSync(fd);
      closeSync(fd);
      fd = null;
      renameSync(tmpPath, filePath);
      syncDirectory(dataDir);
      dirty = false;
      store = merged;
      pending.clear();
    } catch (error) {
      if (fd !== null) closeSync(fd);
      if (existsSync(tmpPath)) unlinkSync(tmpPath);
      throw error;
    } finally {
      releaseLock(dataDir, lockPath, token);
    }
  }

  return { get, set, flush };
}

File: src/cursors.rs
//! Cursor store for mutable byte streams and immutable closed segments, plus
//! the single-writer state lease. Legacy records retain their numeric coercion
//! contract; tagged segment commits are stored as structured records and never
//! projected onto byte-offset fields. Writes are durable: unique temp, file
//! sync, rename, then parent sync.
use std::fs::{self, OpenOptions};
use std::io::Write;
use std::path::{Path, PathBuf};

use serde::Deserialize;
use serde_json::{json, Map, Value};

use crate::util::{mtime_ms, Error, Result};

mod lease;
mod source;
pub use lease::open_writer_lease;
use lease::{acquire_lock, release_lock, sync_directory};
pub use source::SourceCheckpoint;

const SEGMENT_KIND: &str = "closed-segment";

/// A complete-line resume point with proof of the consumed source prefix.
#[derive(Debug, Clone, Copy)]
pub struct ByteCursor {
    pub mtime_ms: f64,
    pub size: u64,
    pub offset: u64,
    pub source: Option<SourceCheckpoint>,
}

impl ByteCursor {
    pub fn is_current(&self, meta: &fs::Metadata) -> bool {
        self.source.is_some_and(|source| source.matches(meta))
            && self.mtime_ms == mtime_ms(meta)
            && self.size == meta.len()
            && self.offset >= meta.len()
    }
}

/// What the store holds for one source path.
#[derive(Debug, Clone)]
pub enum CursorRecord {
    Bytes(ByteCursor),
    /// An immutable closed segment commit, passed through untouched.
    Segment(Value),
}

fn coerce(record: &Value) -> Result<ByteCursor> {
    let number = |key: &str| record.get(key).and_then(Value::as_f64);
    let (Some(mtime_ms), Some(size), Some(offset)) =
        (number("mtimeMs"), number("size"), number("offset"))
    else {
        return Err(Error("cursor record contains invalid numeric state".into()));
    };
    if !mtime_ms.is_finite()
        || !size.is_finite()
        || !offset.is_finite()
        || size < 0.0
        || offset < 0.0
    {
        return Err(Error("cursor record contains invalid numeric state".into()));
    }
    Ok(ByteCursor {
        mtime_ms,
        size: size as u64,
        offset: offset as u64,
        source: Option::<SourceCheckpoint>::deserialize(
            record.get("source").unwrap_or(&Value::Null),
        )?,
    })
}

/// Cursor loss can replay already-persisted evidence, so unreadable state is a
/// hard failure. Recovery uses a separate empty LAKE_DATA root, never a silent
/// fallback that appends a second copy to existing partitions.
fn read_store(path: &Path) -> Result<Map<String, Value>> {
    if !path.exists() {
        return Ok(Map::new());
    }
    let raw = fs::read_to_string(path).map_err(|error| {
        Error(format!(
            "cursor store is unreadable; preserve it and recover into an empty LAKE_DATA: {error}"
        ))
    })?;
    let parsed: Value = serde_json::from_str(&raw).map_err(|error| {
        Error(format!(
            "cursor store is corrupt; preserve it and recover into an empty LAKE_DATA: {error}"
        ))
    })?;
    match parsed {
        Value::Object(map) => Ok(map),
        _ => Err(Error(
            "cursor store must be a JSON object; preserve it and recover into an empty LAKE_DATA"
                .into(),
        )),
    }
}

/// The resume-point store: read at open, merged under its own lock on flush.
pub struct Cursors {
    data_dir: PathBuf,
    file_path: PathBuf,
    lock_path: PathBuf,
    store: Map<String, Value>,
    pending: Map<String, Value>,
    dirty: bool,
}

impl Cursors {
    pub fn open(data_dir: &Path) -> Result<Self> {
        let file_path = data_dir.join("cursors.json");
        let store = read_store(&file_path)?;
        Ok(Self {
            data_dir: data_dir.to_path_buf(),
            file_path,
            lock_path: data_dir.join("cursors.lock"),
            store,
            pending: Map::new(),
            dirty: false,
        })
    }

    pub fn get(&self, file: &str) -> Result<Option<CursorRecord>> {
        let Some(record) = self.store.get(file) else {
            return Ok(None);
        };
        if !record.is_object() {
            return Ok(None);
        }
        if record.get("kind").and_then(Value::as_str) == Some(SEGMENT_KIND) {
            return Ok(Some(CursorRecord::Segment(record.clone())));
        }
        Ok(Some(CursorRecord::Bytes(coerce(record)?)))
    }

    pub fn set_bytes(&mut self, file: &str, cursor: ByteCursor) {
        let value = json!({
            "mtimeMs": cursor.mtime_ms,
            "size": cursor.size,
            "offset": cursor.offset,
            "source": cursor.source,
        });
        self.store.insert(file.to_string(), value.clone());
        self.pending.insert(file.to_string(), value);
        self.dirty = true;
    }

    pub fn set_segment(&mut self, file: &str, record: Value) {
        self.store.insert(file.to_string(), record.clone());
        self.pending.insert(file.to_string(), record);
        self.dirty = true;
    }

    /// Publish pending records. The lock protects the entire read-modify-write
    /// transaction, not merely the final rename.
    pub fn flush(&mut self) -> Result<()> {
        if !self.dirty {
            return Ok(());
        }
        fs::create_dir_all(&self.data_dir)?;
        let token = acquire_lock(&self.data_dir, &self.lock_path)?;
        let outcome = (|| -> Result<Map<String, Value>> {
            let mut merged = read_store(&self.file_path)?;
            for (file, value) in &self.pending {
                merged.insert(file.clone(), value.clone());
            }
            let tmp_path = PathBuf::from(format!(
                "{}.tmp-{}-{}",
                self.file_path.display(),
                std::process::id(),
                uuid::Uuid::new_v4()
            ));
            let write = (|| -> Result<()> {
                let mut file = OpenOptions::new()
                    .write(true)
                    .create_new(true)
                    .open(&tmp_path)?;
                file.write_all(
                    serde_json::to_string_pretty(&Value::Object(merged.clone()))?.as_bytes(),
                )?;
                file.sync_all()?;
                drop(file);
                fs::rename(&tmp_path, &self.file_path)?;
                sync_directory(&self.data_dir)?;
                Ok(())
            })();
            if let Err(error) = write {
                let _ = fs::remove_file(&tmp_path);
                return Err(error);
            }
            Ok(merged)
        })();
        release_lock(&self.data_dir, &self.lock_path, &token);
        let merged = outcome?;
        self.store = merged;
        self.pending.clear();
        self.dirty = false;
        Ok(())
    }
}
